README.md
# splot
An R package to ease data visualization.

The aim of this package is to make visualization an early part of the data analysis process by automating a few common plotting tasks.

In terms of design, it has three general principles:

* **Flexibility**: splot prefers to try to make a reasonable plot from any input, rather than erroring out.
* **Minimal specification**: You should be able to make most plots with just a formula input, and the same formula should generally be compatible with multiple plot types, regardless of variable types.
* **Tweakability**: Though splot is focused on quick, automated plotting, you should be able to adjust any aspect of the display with additional arguments, if you find a plot you want to display elsewhere.

## features
By entering a formula as the first argument in the `splot` function (e.g., `splot(y ~ x)`), you can make

* Density distributions (overlaid on histograms when there is no `by` variable)
* Scatter plots with prediction lines
* Bar or line graphs with error bars

For each type, multiple `y` variables or data at levels of a `by` variable are shown in the same plot frame,<br />
and data at levels of one or two `between` variables are shown in separate plot frames, organized in a grid.

## resources
* [Introduction](https://miserman.github.io/splot/articles/splot.html)
* [Documentation](https://miserman.github.io/splot/reference/splot.html)  
* [Style Guide](https://miserman.github.io/splot/articles/style.html)  
* [Gallery](https://miserman.github.io/splot/articles/gallery.html)
* Applied examples: [Exploring Data](https://miserman.github.io/splot/articles/explore.html) |
[Refining a Result](https://miserman.github.io/splot/articles/refine.html)

## installation
Download R from [r-project.org](https://www.r-project.org/).

Release ([version 0.5.4](https://CRAN.R-project.org/package=splot))
```R
install.packages("splot")
```
Development (version 0.5.5)
```R
# install.packages("remotes")
remotes::install_github("miserman/splot")
```
Then load the package:
```R
library(splot)
```
## examples
Make some data: random group and x variables, and a y variable related to x:
```R
group = rep(c("group 1", "group 2"), 50)
x = rnorm(100)
y = x * .5 + rnorm(100)
```
The distribution of y:
```R
splot(y)
```
A scatter plot between y and x:
```R
splot(y ~ x)
```
Same data with a quadratic model:
```R
splot(y ~ x + x^2 + x^3)
```
Same data separated by group:
```R
splot(y ~ x * group)
```
Could also separate by median or standard deviations of x:
```R
splot(y ~ x * x)
splot(y ~ x * x, split = "sd")
```
Summarize with a bar plot:
```R
splot(y ~ x * group, type = "bar")
```
Two-level y variable with a probability prediction line:
```R
# make some new data for this example:
# a discrete y variable and related x variable:
y_bin = rep(c(1, 5), 50)
x_con = y_bin * .4 + rnorm(100)

# lines = "prob" for a prediction line from a logistic model:
splot(y_bin ~ x_con, lines = "prob")
```


NEWS.md
# splot 0.5.4

## Improvements
* improves export name handling.

# splot 0.5.3

## Additions
* adds splot.colorcontrast.

## Improvements
* improves reference resolution in colors and add.
* allows for more control over graphical parameters.

## Bug fixes
* accounts for NULL stringsAsFactors option.
* appropriately drops levels when they are based on a factor.

# splot 0.5.2

## Additions
* dark option to more easily set text colors for dark backgrounds.

## Improvements
* standardizes color codes before passing them to par.
* avoids overwriting par color settings.
* better handles arguments outside of options when options is specified.
* treats logical y variables as binary rather than factors.

# splot 0.5.1

## Additions
* check equivalency of expression outputs within splot.bench.

## Improvements
* handles input with as.matrix but not as.data.frame methods.
* adds a cutoff option and relative timings to splot.bench for better
  display, and includes more in its return.
* splot.bench now evaluates within the global environment so loaded
  packages can be used.

# splot 0.5.0

## Additions
* specify how colors are interpolated with color.summary.
* calculate average colors with the splot.colormean function.
* fill the area under each density lines with its color with
  density.fill = TRUE (default), adjust the fill's opacity
  with density.opacity, and further control the densities with
  arguments in density.args (which replaces the old bw and adj
  arguments).

## Improvements
* broadens colorby applicability and flexibility.
* tweaks default spacing rules to maximize plot areas.
* aligns axis label specification for density with other types.
* better retains or infers names for multiple variables.
* avoids an error when colorby's by has missing values.

## Bug fixes
* corrects several colorby assignment issues.
* handles mixed-type multiple ys by converting non-numeric variables
  to numeric when other numeric variables are included.
* maintains variable ordering when displaying counts of categorical ys.
* includes colorby variable name in split note when split.

# splot 0.4.2

## Additions
* assign colors in the scale of numeric values with splot.color's
  grade argument.

## Improvements
* makes a legend for colorby in more cases.
* allows renaming and reordering of character or factor y levels.
* more broadly retains character and factor level orders.
* more broadly accounts for missing values.
* better handles colorby x aggregation.
* splot.color returns shorter color codes when opacity is 1.

## Bug fixes
* corrects application of adjustments in splot.color when no sampling
  method is used.
* corrects connected line paths when sparse.
* calculates frequencies of y levels within by and between splits.
* corrects some colorby x and by assignments.
* preserves color assignments when colorby has multiple arguments,
  and some levels of by are empty.
* avoids error when a level of split is empty.

# splot 0.4.1

## Improvements
* allows splot.color sampling to be turned off.
* allows splot.bench to handle single runs.
* enables more sorting options.

## Bug fixes
* correctly does not display error bars when error is set to FALSE.
* corrects the x-axis range of scatter plots when x is logical.
* corrects bet reassignment given multiple ys.

# splot 0.4.0

## Additions
* place the legend interactively.
* adjust color opacity, and assign colors in gradients.
* split by specified values.
* display predicted probability lines in scatter plots with 2-level ys.
* allows for non-numeric y variables.

## Improvements
* general improvements in automatic settings.
* more reliable connection between data and line colors, types, and widths.
* provides more control over the legend and colorby.
* improved reading of the y and su arguments.
* improved bar plot y-axis labels.

## Bug fixes
* now appropriately recycles short variables.
* avoids an error when levels of by are missing within between splits.

# splot 0.3.2

## Additions
* sort text x variables by their y values.
* split by a specified number of segments.

## Improvements
* cleared up sub/sud related mixups.
* level ordering now applies to between variables.
* improved formula parsing; expressions should be better recognized
  as arguments.
* rearranged arguments to be more like other functions; variables other
  that y must be named, or entered as part of the y formula.
* multiple variables entered as y can now be renamed/ordered with an mv entry
  in levels.

## Bug fixes
* corrected y-axis labeling for bar plots when autori is true.
* custom notes now overwrite all other note elements.
* specified x levels are now appropriately applied when mv.as.x is true.
* standardized lty behavior across types, and prevented it from causing errors.

tests/testthat/test-splot.R
n <- 2000
d <- data.frame(sapply(c("c1", "c2", "c3"), function(c) sample(0:1, n, TRUE)))
d$v1 <- with(
  d,
  rnorm(n) + c1 * -.4 + c1 * c2 * -.3 + c1 * c3 *
    .3 + c2 * c3 * .9 - .8 + rnorm(n, 0, c1)
)
d$v2 <- with(
  d,
  v1 * .2 + c1 * .3 + c3 * -.6 + c2 * c3 * .8 + v1 *
    c1 * c2 * -.5 + v1 * c1 * c2 * c3 * -.5
    + rnorm(n, 5) + rnorm(n, -1, .1 * v1^2)
)

test_that("density works", {
  expect_warning(splot(d$v1, type = "scatter"), "x must be included")
  expect_true(splot(v1 ~ c1, d, type = "d")$ck$t == 2)

  res <- splot(d$v1)
  expect_true(res$ck$t == 2)
  expect_identical(res$cdat$`.^^.`$y, d$v1)
  expect_identical(res$dat, splot(v1, d)$dat)

  res <- splot(v1, d, by = c1)
  expect_true(res$ck$t == 2)
  expect_true(res$lega$title == "c1")

  res <- splot(v1, d, by = c1, between = c2)
  expect_true(res$ck$t == 2)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c2)))

  res <- splot(v1, d, by = c1, between = c(c2, c3))
  expect_true(res$ck$t == 2)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c2, d$c3)))
})

test_that("line/bar works", {
  expect_true(splot(v1 ~ c1, d, type = "d")$ck$t == 2)

  res <- splot(d$v1 ~ d$v2, type = "l")
  expect_true(res$ck$t == 1)
  expect_identical(unname(res$ptxt$l.x), c("Under Median", "Over Median"))
  expect_identical(
    unname(splot(d$v1 ~ d$v2, type = "b", split = "mean")$ptxt$l.x),
    c("Below Average", "Above Average")
  )

  res <- splot(d$v1 ~ d$c1)
  expect_true(res$ck$t == 1)
  expect_identical(res$cdat$`.^^.`, data.frame(y = d$v1, x = as.factor(d$c1)))
  expect_identical(res$dat, splot(v1 ~ c1, d)$dat)

  res <- splot(v1 ~ c1, d, by = c2, model = TRUE)
  expect_identical(names(res$fmod$coefficients), c("(Intercept)", "c1", "c2", "c1:c2"))
  expect_true(res$ck$t == 1)
  expect_true(res$lega$title == "c2")
  expect_identical(res$dat, splot(v1 ~ c1 * c2, d)$dat)

  res <- splot(v1 ~ c1, d, by = c2, between = c3)
  expect_true(res$ck$t == 1)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c3)))
  expect_identical(res$dat, splot(v1 ~ c1 * c2 * c3, d)$dat)

  res <- splot(v1 ~ v2, d, by = "c1", between = c("c2", "c3"), type = "b")
  expect_true(res$ck$t == 1)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c2, d$c3)))
  expect_identical(res$dat, splot(v1 ~ v2 * c1 * c2 * c3, d, type = "l")$dat)
})

test_that("scatter works", {
  expect_true(splot(v1 ~ c1, d, type = "s")$ck$t == 3)

  res <- splot(d$v1 ~ d$v2)

  res <- splot(v1 ~ v2, d, by = c1, model = TRUE)
  expect_identical(names(res$fmod$coefficients), c("(Intercept)", "v2", "c1", "v2:c1"))
  expect_true(res$ck$t == 3)
  expect_true(res$lega$title == "c1")
  expect_true(res$ck$t == 3)
  expect_identical(
    unname(splot(d$v1 ~ d$v2 * d$v1, split = "mean")$ptxt$l.by),
    c("Below Average", "Above Average")
  )
  expect_identical(res$dat, splot(v1 ~ v2 * c1, d)$dat)

  res <- splot(v1 ~ v2, d, by = "c1", between = "c2")
  expect_true(res$ck$t == 3)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c2)))
  expect_identical(res$dat, splot(v1 ~ v2 * c1 * c2, d)$dat)

  res <- splot(v1 ~ v2, d, by = c1, between = c(c2, c3))
  expect_true(res$ck$t == 3)
  expect_identical(as.numeric(res$seg$n), as.numeric(table(d$c2, d$c3)))
  expect_identical(res$dat, splot(v1 ~ v2 * c1 * c2 * c3, d)$dat)
})

test_that("colorby works", {
  res <- splot(d$v1 ~ d$v2, colorby = .y)
  expect_true(res$ptxt$cbo == ".y")
  expect_identical(res$ptxt$leg, splot(d$v1 ~ d$v2, colorby = d$v1)$ptxt$leg)
  expect_identical(
    sort(splot(as.character(c1), d, colorby = v1)$lega$legend),
    sort(unname(formatC(tapply(d$v1, d$c1, mean), 2, format = "f")))
  )
})

skip_if_not(capabilities("cairo"), "Cairo devices not supported")
test_that("saving works", {
  file <- paste0(tempfile(), ".svg")
  splot(1:10, file.name = file, format = "svg")
  expect_true(file.exists(file))
})


tests/testthat/test-splot.bench.R
test_that("basic example work", {
  res <- splot.bench(Sys.sleep(.01), Sys.sleep(.1), runsize = 1)
  expect_true(res$summary[3, 1] < res$summary[3, 2])
})


tests/testthat/test-splot.color.R
base <- c(
  "#82C473", "#A378C0", "#616161", "#9F5C61", "#D3D280", "#6970B2", "#78C4C2", "#454744", "#D98C82"
)

test_that("base cases work", {
  expect_identical(splot.color(), base)
  expect_true(!all(splot.color(rep(1, 5), method = "related") == splot.color(rep(1, 5))))
  expect_identical(splot.color(c(1, 1), by = 1:2), base[1:2])
  expect_identical(splot.color(c(1, 1), by = 1:2, flat = FALSE), list(`1` = base[1], `2` = base[2]))
  expect_identical(
    unname(splot.color(list(c(1, 1), c(1, 1)), method = "none")),
    rep(base[1:2], each = 2)
  )
})

test_that("scale works", {
  initial <- splot.color(1:5)
  expect_identical(order(-colMeans(col2rgb(initial))), seq_len(5))
  expect_identical(
    order(-colMeans(col2rgb(initial))),
    order(colMeans(col2rgb(splot.color(1:5, decreasing = TRUE))))
  )
  expect_true(all(initial %in% splot.color(1:5, shuffle = TRUE)))
})


R/splot.colormean.R
#' splot color average
#'
#' Calculates the average of a set of colors, returning its Hex code.
#' @param ... color codes or names as characters.
#' @return The calculated color code.
#' @examples
#' # average of red and blue
#' plot(
#'   1:3, numeric(3),
#'   pch = 15, cex = 20, xlim = c(0, 4),
#'   col = c("red", splot.colormean("red", "blue"), "blue")
#' )
#'
#' # average of a set
#' x <- rnorm(100)
#' set <- splot.color(x, method = "related")
#' splot(
#'   x ~ rnorm(100),
#'   colors = set,
#'   add = points(0, 0, pch = 15, cex = 10, col = splot.colormean(set))
#' )
#' @export

splot.colormean <- function(...) {
  hdc <- c(0:9, LETTERS[1:6])
  hdc <- outer(hdc, hdc, paste0)
  s <- seq_len(16)
  ccs <- adjustcolor(unlist(list(...), use.names = FALSE))
  paste(c("#", apply(Reduce("+", lapply(ccs, function(cc) {
    cc <- strsplit(cc, "")[[1]][2:7]
    cc <- paste0(cc[c(TRUE, FALSE)], cc[c(FALSE, TRUE)])
    vapply(cc, function(c) which(hdc == c, TRUE), numeric(2))
  })) / length(ccs), 2, function(cc) {
    hdc[which.min(abs(s - cc[1])), which.min(abs(s - cc[2]))]
  })), collapse = "")
}


R/utils.R
#' splot colors
#'
#' Get a prespecified set of 9 colors, or a set of graded or random, potentially grouped colors.
#' @param x dictates the number and shade of colors. If a single value, returns that many samples of the
#'   first \code{seed} entry. If a vector, returns a color for each entry. If numeric, a single seed color
#'   is sampled in order of the vector. If a character or factor, a separate seed color is assigned to
#'   each level, then sampled within levels. Values or vectors in a list are each assigned a seed color.
#' @param by a vector to group \code{x} by; each level is assigned a seed color.
#' @param seed a vector of color names or codes to adjust from, lining up with levels of \code{x} or
#'   \code{by}, or the name of a palette, partially matching \code{'bright'}, \code{'dark'},
#'   \code{'pastel'}, or \code{'grey'}.
#' @param brightness adjusts the RGB values of the seed color, usually between -1 and 1.
#' @param luminance adjusts the white levels of the seed color, usually between -1 and 1.
#' @param opacity sets the opacity of the seed color, between 0 and 1.
#' @param extend if \code{method='scale'}, extends the range of the gradient beyond the sampled range,
#'   making for more similar colors (defaults is .5, with 0 sampling the full range). If
#'   \code{method='related'}, increases the amount any of the RGB values can be adjusted, making for
#'   potentially more different colors (default is 2).
#' @param lighten logical; if \code{TRUE}, scaled colors are lightened instead of darkened. Only
#'   applicable if \code{method='scale'}.
#' @param shuffle logical; if \code{TRUE}, scaled colors are shuffled. Only applicable if
#'   \code{method='scale'}.
#' @param flat logical; if \code{FALSE} and \code{x} is a character, factor, or list, or \code{by} is not
#'   missing, a list is returned.
#' @param method a character setting the sampling method: If \code{'related'} (\code{'^rel|^ran|^o'}),
#'   RGB values are freely adjusted, resulting in similar colors. If \code{'none'} (\code{'^no|^f|^bin'}),
#'   Seed colors are simply repeated in each level (sampling is off). Otherwise, RGB values are adjusted
#'   together, resulting in a gradient.
#' @param grade logical; if \code{TRUE}, seeds are adjusted on the scale of numeric \code{x}s.
#'   Otherwise, seeds are adjusted in even steps along numeric \code{x}s.
#' @param decreasing logical; if \code{FALSE}, assigns colors to numeric \code{x}s in increasing order.
#' @param nas value to replace missing values with.
#' @details
#' If \code{x} and \code{by} are not specified (or are characters with a length of 1, in which case they
#' are treated as \code{seed}), only the seed palette is returned.
#'
#' To expand on a palette, seed colors are assigned to groups, and variants of each seed are assigned to
#' values or levels within groups, or randomly or as a gradient if there are no values or level to assign to.
#'
#' Seed colors are assigned to groups. If \code{x} is a character or factor and no \code{by} has been
#' specified, groups are the unique levels of \code{x}. If \code{by} is specified and is a character or
#' factor, or has fewer than 10 unique levels, groups are levels of \code{by}. If \code{x} is a list,
#' groups are list entries.
#'
#' The number of variants for each seed color is determined either by a value (if the value has a length
#' of 1; e.g., \code{x=10}), the vector's length (if \code{x} is numeric), or the count of the given level
#' (if \code{x} is a factor or character vector).
#'
#' @examples
#' # including no arguments or just a palette name will only return
#' # the palette as a character vector
#' pastel_palette = splot.color()
#' dark_palette = splot.color('dark')
#'
#' # entering a number for x will generate that many variants of the first seed color
#' red_scale = splot.color(10, 'red')
#'
#' # entering a list of values as x will return that many variants of the associated seed
#' red_and_green_scales = splot.color(list(10,10), seed=c('red','green'))
#'
#' # this shows gradients of each color in the default palette
#' # a list entered as colorby is treated as arguments to splot.color
#' # periods before the position name refer to the internally assembled data
#' splot(
#'   rep(splot.color(),each=100)~rep.int(seq.int(.01,1,.01),9),colorby=list(.x,.y),
#'   lines=FALSE,mar=c(2,4,0,0),cex=c(points=3),leg=FALSE,pch=15,
#'   title="'pastel' palette",labx='value of x',laby='seed color'
#' )
#'
#' # colors graded by value, entered in a list
#' plot(1:30,numeric(30),pch=15,cex=10,col=splot.color(list(1:8,c(7:1,1:7),8:1)))
#'
#' # comparing sampling methods:
#' #   on top are 1000 similar colors, with different RGB ratios
#' #   on bottom are 268 colors with the same RGB ratio at different levels
#' splot(
#'   c(rnorm(1000),rnorm(1000,10))~rnorm(2000),lines=FALSE,
#'   colors=c(splot.color(1000),splot.color(1000,method='related'))
#' )
#'
#' @export

splot.color=function(x=NULL,by=NULL,seed='pastel',brightness=0,luminance=0,opacity=1,extend=.7,
  lighten=FALSE,shuffle=FALSE,flat=TRUE,method='scale',grade=FALSE,decreasing=FALSE,nas='#000000'){
  sets=list(
    bright=c('#45FF00','#BA00FF','#000000','#FF0000','#FFFD00','#003DFF','#00F2F8','#999999','#FF891B'),
    dark=c('#1B8621','#681686','#2A2A2A','#7C0D0D','#B5BC00','#241C80','#1A7E8B','#666666','#B06622'),
    pastel=c('#82C473','#A378C0','#616161','#9F5C61','#D3D280','#6970B2','#78C4C2','#454744','#D98C82'),
    grey=function(n) grey(.2:n/(n+n*if(n<10) .1 else .3))
  )
  if(missing(seed) && is.character(x) && (length(x) == 1 || all(tolower(x) %in% colors()))){
    seed = x
    x = numeric(length(seed)) + 1
  }
  if(missing(seed) && is.character(by) && length(by)==1){
    seed=by
    by=NULL
  }
  seed=tolower(seed)
  ox=NULL
  lvs=function(x) if(is.factor(x)) base::levels(x) else sort(unique(x[!is.na(x)]))
  cn = ncol(x)
  if(!is.null(cn) && !is.na(cn) && cn > 1){
    if(is.null(by)) by=x[,2]
    x=x[,1]
  }else if(is.list(x) && length(x)==1) x=x[[1]]
  if(!is.null(by)){
    ol=length(x)
    if(is.null(x)){
      x=by
      by=NULL
    }else if(ol!=length(by)){
      if(is.numeric(by) && length(by)==1 && by<ol) by=rep_len(seq_len(by),ol) else{
        by=NULL
        warning('splot.color: by was dropped as it is not the same length as x', call. = FALSE)
      }
    }
  }
  if(!is.null(x) && (!(is.list(x) || is.numeric(x)) || (is.numeric(x) && !is.null(by)))){
    if(is.null(by)){
      ox=x
      x=as.list(table(x))[lvs(ox)]
    }else{
      if(is.numeric(by) && length(lvs(by))>9) warning('splot.color: only non-numeric bys are accepted', call. = FALSE) else{
        ox = by = factor(by, lvs(by))
        x = split(x, by)
      }
    }
  }
  ol=length(x)
  if(ol==1 && is.list(x)){
    x=x[[1]]
    ol=length(x)
    ox=NULL
  }
  n=if(ol==1) x else ol
  if(length(seed)==1 && grepl('^bri|^dar|^pas|^gr[ae]y',seed)){
    seed=match.arg(seed,names(sets))
    seed=if(seed=='grey') if(n==1) '#666666' else if(ol==1)
      return(sets$grey(n)) else sets$grey(n) else sets[[seed]]
    if(is.null(x) || (ol==1 && n<2)) return(seed)
  }
  ckno = grepl('^no|^f|^bin', method, TRUE)
  sc=if(grepl('^rel|^ran|^o',method,TRUE)){
    r=if(missing(extend)) 2 else max(.001,extend)
    function(cc,n){
      cc=adjustcolor(cc)
      hdc=c(0:9,LETTERS[1:6])
      hdc=outer(hdc,hdc,paste0)
      ccord=function(cc){
        cc=strsplit(cc,'')[[1]][2:7]
        cc=paste0(cc[c(TRUE,FALSE)],cc[c(FALSE,TRUE)])
        vapply(cc,function(c)which(hdc==c,TRUE),c(0,0))
      }
      ccode=function(m){
        s=seq_len(16)
        paste0('#',paste(apply(m,2,function(cc){
          hdc[which.min(abs(s-cc[1])),which.min(abs(s-cc[2]))]
        }),collapse=''))
      }
      csamp=function(code,n){
        n=max(1,n-1)
        ocs=NULL
        code=ccord(code)
        if(any(ck<-code>14)) code[ck]=code[ck]-(code[ck]-14)
        if(any(ck<-code<2)) code[ck]=code[ck]+(2-code[ck])
        i=1
        while(length(ocs)<=n && i<9999){
          s=sample(1:6,3)
          nc=code
          nc[s]=nc[s]+sample(-r:r,3,TRUE)
          nc=ccode(nc)
          if(!nc%in%ocs) ocs=c(ocs,nc)
          i=i+1
        }
        if(any(opacity!=1,brightness!=0,luminance!=0)){
          adj=1+brightness
          ocs=adjustcolor(ocs,opacity,adj,adj,adj,c(rep(luminance,3),0))
        }
        if(length(ocs)!=n+1) ocs=rep_len(ocs,n+1)
        ocs
      }
      csamp(cc,n)
    }
  }else if(ckno) function(cc, n){
    ns = length(n)
    vapply(seq_len(ns), function(i){
      adj = ns / (ns + i - 1) + brightness
      if(lighten) adj = 1.8 - adj
      adjustcolor(cc, opacity, adj, adj, adj, c(rep(luminance, 3), 0))
    }, '')
  }else function(cc, n){
    s = abs(n - max(n, na.rm = TRUE))
    n = length(s)
    s = s / max(s, na.rm = TRUE) * (n - 1) + 1
    r = max(n, n + n * extend)
    if(!lighten) s = s + r - max(s, na.rm = TRUE)
    vapply(s, function(i){
      adj = i / r + brightness
      if(lighten) adj = adj + 1
      adjustcolor(cc, opacity, adj, adj, adj, c(rep(luminance, 3), 0))
    }, '')
  }
  asc = function(v, si){
    n = length(v)
    if(is.numeric(v)) v = round(v, 3)
    if(!is.numeric(v) || (n != 1 && !grade)) v = as.numeric(factor(v, lvs(v)))
    if(!ckno) if(n == 1) v = seq_len(max(1, v)) else if(length(lvs(v)) == 1) v = seq_along(v)
    n = length(v)
    l = lvs(v)
    u = sort(unique(v), decreasing)
    nu = length(u)
    pr = rep(nas, n)
    cols = if(nu < 2 && (!length(u) || u < 2)) si else sc(si, u)
    v = factor(v[is.finite(v)], u)
    if(n != nu) cols = rep(cols, tabulate(v))
    if(shuffle) sample(cols) else cols[order(order(v, decreasing = decreasing))]
  }
  if(!is.list(x)){
    seed = asc(x, seed[1])
  }else{
    if(length(seed) < n) seed = rep_len(seed, n)
    seed = lapply(seq_len(n), function(i) asc(x[[i]], seed[i]))
    names(seed)=if(!is.null(names(x))) names(x) else vapply(seed,'[[','',1)
    if(flat) seed=if(!is.null(ox) && all(lvs(ox)%in%names(seed))){
      by=as.character(ox)
      for(g in lvs(ox)){
        su = !is.na(by) & !is.nan(by) & by == g
        ssu = sum(su)
        if(is.finite(ssu) && ssu) by[su] = rep_len(seed[[g]], ssu)
      }
      by
    }else unlist(seed)
  }
  if(!is.list(seed)) seed[is.na(seed) | is.nan(seed) | seed %in% c('NA', 'NaN', 'Inf', '-Inf')] = nas
  if(opacity == 1) seed = if(is.list(seed)) lapply(seed, function(s) sub('FF$', '', s)) else sub('FF$', '', seed)
  seed
}

#' splot color average
#'
#' Calculates the average of a set of colors, returning its Hex code.
#' @param ... color codes or names as characters.
#'
#' @examples
#' # average of red and blue
#' plot(
#'   1:3, numeric(3), pch = 15, cex = 20, xlim = c(0, 4),
#'   col = c('red', splot.colormean('red', 'blue'), 'blue')
#' )
#'
#' # average of a set
#' x = rnorm(100)
#' set = splot.color(x, method = 'related')
#' splot(
#'   x ~ rnorm(100), colors = set,
#'   add = points(0, 0, pch = 15, cex = 10, col = splot.colormean(set))
#' )
#'
#' @export

splot.colormean = function(...){
  hdc = c(0:9, LETTERS[1:6])
  hdc = outer(hdc, hdc, paste0)
  s = seq_len(16)
  ccs = adjustcolor(unlist(list(...), use.names = FALSE))
  paste(c('#', apply(Reduce('+', lapply(ccs, function(cc){
    cc = strsplit(cc, '')[[1]][2:7]
    cc = paste0(cc[c(TRUE, FALSE)], cc[c(FALSE, TRUE)])
    vapply(cc, function(c) which(hdc == c, TRUE), numeric(2))
  })) / length(ccs), 2, function(cc){
    hdc[which.min(abs(s - cc[1])), which.min(abs(s - cc[2]))]
  })), collapse = '')
}

#' splot benchmarker
#'
#' Time one or more expressions over several iteration, then plot the distributions of their times.
#' @param ... accepts any number of expressions to be timed. See examples.
#' @param runs the number of overall iterations. Increase to stabilize estimates.
#' @param runsize the number of times each expression is evaluated within each run. Increase to
#'   differentiate estimates (particularly for very fast operations).
#' @param cleanup logical; if \code{TRUE}, garbage collection will be performed before each run.
#'   Garbage collection greatly increases run time, but may result in more stable timings.
#' @param print.names logical; if \code{FALSE}, the entered expressions will be included in the plot
#'   as legend names. Otherwise, (and if the number of expressions is over 5 or the length of any
#'   expression is over 50 characters) expressions are replaced with numbers corresponding to their
#'   entered position.
#' @param limit.outliers logical; if \code{TRUE} (default), times over an upper bound for the given
#'   expression will be set to that upper bound, removing aberrant extremes.
#' @param check_output logical; if \code{TRUE}, the output of each expression is checked with
#'   \code{\link[base]{all.equal}} against that of the first. A warning indicates if any are not
#'   equal, and results are invisibly returned.
#' @param check_args a list of arguments to be passed to \code{\link[base]{all.equal}}, if
#'   \code{check_output} is \code{TRUE}.
#' @param options a list of options to pass on to splot.
#' @return A list:
#' \tabular{ll}{
#'   plot \tab splot output\cr
#'   checks \tab a list of result from all.equal, if \code{check_output} was \code{TRUE}\cr
#'   expressions \tab a list of the entered expressions \cr
#'   summary \tab a matrix of the printed results \cr
#' }
#' @examples
#' # increase the number of runs for more stable estimates
#'
#' # compare ways of looping through a vector
#' splot.bench(
#'   sapply(1:100, '*', 10),
#'   mapply('*', 1:100, 10),
#'   vapply(1:100, '*', 0, 10),
#'   unlist(lapply(1:100, '*', 10)),
#'   runs = 20, runsize = 200, check_output = TRUE
#' )
#'
#' # compare ways of setting all but the maximum value of each row in a matrix to 0
#' \dontrun{
#'
#' mat = matrix(c(rep(1, 4), rep(0, 8)), 4, 3)
#' splot.bench(
#'   t(vapply(seq_len(4), function(r){
#'     mat[r, mat[r,] < max(mat[r,])] = 0
#'     mat[r,]
#'   }, numeric(ncol(mat)))),
#'   do.call(rbind, lapply(seq_len(4), function(r){
#'     mat[r, mat[r,] < max(mat[r,])] = 0
#'     mat[r,]
#'   })),
#'   do.call(rbind, lapply(seq_len(4), function(r){
#'     nr = mat[r,]
#'     nr[nr < max(nr)] = 0
#'     nr
#'   })),
#'   {nm = mat; for(r in seq_len(4)){
#'     nr = nm[r,]
#'     nm[r, nr < max(nr)] = 0
#'   }; nm},
#'   {nm = mat; for(r in seq_len(4)) nm[r, nm[r,] < max(nm[r,])] = 0; nm},
#'   {nm = matrix(0, dim(mat)[1], dim(mat)[2]); for(r in seq_len(4)){
#'     m = which.max(mat[r,])
#'     nm[r, m] = mat[r, m]
#'   }; nm},
#'   {ck = do.call(rbind, lapply(seq_len(4), function(r){
#'     nr = mat[r,]
#'     nr < max(nr)
#'   })); nm = mat; nm[ck] = 0; nm},
#'   t(apply(mat, 1, function(r){
#'     r[r < max(r)] = 0
#'     r
#'   })),
#'   runs = 50, runsize = 200, check_output = TRUE
#' )
#' }
#' @export

splot.bench = function(..., runs = 20, runsize = 200, cleanup = FALSE, print.names = FALSE,
  limit.outliers = TRUE, check_output = TRUE, check_args = list(), options = list()){
  e = sapply(as.character(substitute(list(...)))[-1], function(t) parse(text = t))
  e = e[!duplicated(names(e))]
  es = length(e)
  if(!es) stop('no expressions found', call. = FALSE)
  ne = names(e)
  seconds = matrix(NA, runs, es, dimnames = list(NULL, ne))
  rs = seq_len(runsize)
  ops = tryCatch(
    lapply(e, eval, .GlobalEnv),
    error = function(e) stop('one of your expressions breaks:\n', e, call. = FALSE)
  )
  checks = if(check_output && length(e) != 1){
    if(!'check.attributes' %in% names(check_args)) check_args$check.attributes = FALSE
    if(!'check.names' %in% names(check_args)) check_args$check.names = FALSE
    lapply(ops[-1], function(r) tryCatch(
      do.call(all.equal, c(list(r), list(ops[[1]]), check_args)),
      error = function(e) FALSE
    ))
  }else NULL
  if(!is.null(checks) && !all(vapply(checks, isTRUE, TRUE))) warning(
    'some of your expressions do not seem to have similar results as the first;',
    ' see the `checks` output.', call. = FALSE
  )
  ost = proc.time()[3]
  cat('benchmarking', es, 'expression(s) in chunks of', runsize, 'per run... \nrun 0 of', runs)
  fun = function(e){
    eval(e, .GlobalEnv)
    NULL
  }
  for(r in seq_len(runs)){
    for(f in sample(seq_len(es))){
      if(cleanup) gc(FALSE)
      st = proc.time()[[3]]
      for(i in rs) fun(e[[f]])
      seconds[r, f] = proc.time()[[3]] - st
    }
    cat('\rrun', r, 'of', runs)
  }
  cat('\rfinished', runs, 'runs in', round(proc.time()[3] - ost, 2), 'seconds       \n\n')
  cat('expressions:\n\n')
  icn = seq_len(es)
  ne = gsub('\n', '\n   ', ne, fixed = TRUE)
  for(i in icn) cat(i, '. ', ne[i], '\n', sep = '')
  cat('\n')
  res = rbind(colSums(seconds), colMeans(seconds))
  res = rbind(res, if(min(res[1,], na.rm = TRUE) == 0) res[1,] + 1 else res[1,] / min(res[1,], na.rm = TRUE))
  dimnames(res) = list(c('total time (seconds)', 'mean time per run', 'times the minimum'), icn)
  print(round(res, 4))
  if(!print.names) if(!missing(print.names) || es > 5 || any(nchar(names(e)) > 50))
    colnames(seconds) = icn
  if(limit.outliers) for(f in seq_len(es)){
    qr = quantile(seconds[, f], c(.25, .75), TRUE)
    qrc = qr[2] + (qr[2] - qr[1]) * 1.5
    seconds[seconds[, f] > qrc, f] = qrc
  }
  if(es == 1 && runs == 1) return(list(plot = NULL, summary = res))
  title = paste('timing of', runs, 'runs of', runsize, 'calls each')
  if(nrow(seconds) == 1){
    options$x = colnames(seconds)
    seconds = seconds[1,]
  }
  invisible(list(
    plot = splot(seconds, title = title, labels.filter = FALSE, labels.trim = FALSE, options = options),
    checks = checks,
    expressions = as.list(unname(e)),
    summary = res
  ))
}